### PR TITLE
Creating AWS S3 Config Source

### PR DESCRIPTION
**Pull request by Aditya-Gollapudi**
Moved this from collector to contrib

Note: Please don't review anything in the configprovider folder - that will all be merged in a different PR (coming from Paolo) The only stuff from this that will be merged is in the s3configsource folder. This PR was mostly submitted to try to parallelize some of the reviewing

Description: Added a Config Source for S3 inline with the new config source changes that have been made

Testing: So far there is only a core test, but I will continue to add them - this is something of a draft. Also please note that we expect the configprovider folder to be merged in at some point later on but that won't live here - where exactly everything is to be merged is still up in the air

@pjanotti

Link to old PR: https://github.com/open-telemetry/opentelemetry-collector/pull/3738

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Please fix the build:

```
# github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter
exporter\awsemfexporter\grouped_metric.go:20:2: json redeclared as imported package name
	previous declaration at exporter\awsemfexporter\grouped_metric.go:18:2
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/cmd/otelcontribcol [build failed]
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/internal/version	0.040s [no tests to run]
FAIL
```

File: internal/aws/awss3configsource/config.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awss3configsource
+
+import (
+	splunkprovider "github.com/signalfx/splunk-otel-collector"
```
Comment (anuraaga) [inline internal/aws/awss3configsource/config.go:18]: It seems quite odd for the upstream collector to depend on a downstream distro, it's like a circular dependency

**Review by anuraaga — COMMENTED** [review]
Is it possible to use an abstraction like vfs to support all clouds?

https://github.com/C2FO/vfs

Reading config files doesn't seem like an extremely high-scale operation or anything that might benefit performance-wise from being "closer to metal" using the raw cloud sdks so maybe it's practical.

File: internal/aws/awss3configsource/go.mod [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module example.com/awss3configsource
+
+go 1.16
+
+require (
+	github.com/aws/aws-sdk-go v1.40.5
+	github.com/knadh/koanf v1.2.0
+	github.com/signalfx/splunk-otel-collector v0.0.0-00010101000000-000000000000
+	github.com/stretchr/testify v1.7.0
+	go.opentelemetry.io/collector v0.31.0
```
Comment (bogdandrutu) [inline internal/aws/awss3configsource/go.mod:10]: Please use the same version as other `go.mod` in the repo

File: internal/aws/awss3configsource/config.go (hunk 1 above)
Comment (Aditya-Gollapudi) [reply internal/aws/awss3configsource/config.go:18]: Splunk is currently submitting a string of PRs into Collector - the plan is to have this merged in - in fact in the go.mod file I replace this reference with a local reference to the config provider folder

**Comment by Aditya-Gollapudi** [comment]
> https://github.com/C2FO/vfs

@anuraaga Is VFS approved by security - I didn't spend a bunch of time thinking about it as I was under the impression that it would be rejected by AppSec out of hand

Additionally, the goal of the config source interface is for each of the cloud providers to eventually add their own config source

**Comment by anuraaga** [comment]
@Aditya-Gollapudi Sorry for missing the comment from 20 days ago just saw it now. In principle, basically nothing is approved by AWS security as they check at a micro level - even if using the AWS SDK, the usage of the AWS SDK needs to get checked, etc. On the flip side, if using a different library like vfs, then the usage of that library would be checked in a similar way. Since VFS itself just wraps the AWS SDK, this will be straightforward.

Also, this is the OpenTelemetry Collector - what's important is creating a component that makes sense for this codebase. If it's possible to make a generic mechanism for configuration across cloud providers, then that is a much better fit for this codebase than another mechanism that duplicates logic, with greater maintenance cost. If we think such a mechanism is simple, secure, and performant, than internal security review can be made to happen.

**Comment by Aditya-Gollapudi** [comment]
@anuraaga Thanks for getting back - I believe that AppSec (perhaps AWS security is different) did approve the usage of the SDK in the way I did above - I didn't know that it was possible to get VFS approved. If that is the case I agree it makes some sense to use it as it more generic.

However, I am no longer at Amazon for the moment so I'm not actively working on this. I believe Tianduo Hao is continuing my work (for some reason I can't seem to @skyduo)

**Pull request closed without merging**